src/flamenco/runtime/fd_executor.c

```diff
@@ -294,71 +294,160 @@ fd_executor_check_txn_program_accounts_and_data_sz( fd_exec_txn_ctx_t * txn_ctx
+        writable try to collect rent from the account. 
+        
+     After the account is loaded accumulate the data size to make sure the
+     transaction doesn't violate the transaction loading limit.
+
+     In the firedancer client only some of these steps are necessary because
+     all of the accounts are loaded in from the accounts db into borrowed
+     accounts already. The instruction sysvar gets loaded in later in 
+     fd_execute_txn
+     1. If the account is writable, try to collect fees on the account. Unlike
+        the agave client, this is also done on the fee payer account. The agave
+        client tries to collect rent on the fee payer while the fee is being
+        collected in validate_fees().
+     2. If the account is not writable and it is not an instruction account
+        and would be in the loaded program cache, then it should be replaced
+        with a dummy value. 
+     */
+
+  fd_epoch_schedule_t const * schedule = &fd_exec_epoch_ctx_epoch_bank( txn_ctx->slot_ctx->epoch_ctx )->epoch_schedule;
+  ulong off;
+  ulong epoch = fd_slot_to_epoch( schedule, txn_ctx->slot_ctx->slot_bank.slot, &off );
 
-  // https://github.com/anza-xyz/agave/blob/eff8961dcf7880aa7e438463203611a8a0088554/svm/src/account_loader.rs#L315-L335
   for( ulong i=0UL; i<txn_ctx->accounts_cnt; i++ ) {
-    // Check for max loaded acct size
     fd_borrowed_account_t * acct = NULL;
-    int err        = fd_txn_borrowed_account_view_idx( txn_ctx, (uchar)i, &acct );
+  
+    int   err      = fd_txn_borrowed_account_view_idx( txn_ctx, (uchar)i, &acct );
     ulong acc_size = err==FD_ACC_MGR_SUCCESS ? acct->const_meta->dlen : 0UL;
 
-    err = accumulate_and_check_loaded_account_data_size( acc_size, requested_loaded_accounts_data_size, &accumulated_account_size );
+    /* Try to collect rent on all writable accounts. If rent is collected
+       successfully, update the starting lamports accordingly to avoid
+       unbalanced lamports issues during instruction execution. 
+       TODO: the rent epoch check in the conditional should probably be moved
+       to inside fd_runtime_collect_rent_account. */
+    if( fd_txn_account_is_writable_idx( txn_ctx, (int)i ) && acct->const_meta->info.rent_epoch<=epoch ) {
+      fd_runtime_collect_rent_account( txn_ctx->slot_ctx, acct->meta, acct->pubkey, epoch );
+      acct->starting_lamports = acct->meta->info.lamports;
+      (void)epoch;
+    }
+
+    err = accumulate_and_check_loaded_account_data_size( acc_size,
+                                                         requested_loaded_accounts_data_size,
+                                                         &accumulated_account_size );
+
     if( FD_UNLIKELY( err!=FD_RUNTIME_EXECUTE_SUCCESS ) ) {
       return err;
     }
   }
 
-  // https://github.com/anza-xyz/agave/blob/ae18213c19ea5335dfc75e6b6116def0f0910aff/svm/src/account_loader.rs#L377-L433
+  fd_pubkey_t program_owners[64];
```

Comment:
`program_owners_cnt` seems to be bounded by `instr_cnt` so you can save some memory initializing like `fd_pubkey_t program_owners[instr_cnt];`